### DISCUSSION TITLE
Mysql service never comes up healthy in action

### DISCUSSION THREAD

**Discussion by peburrows** [community/community#27021] [Actions]
<p>At some point in the last couple weeks, the mysql:5.7 service config I was using to expose mysql to my action stopped working. Here’s the config I’m using:</p>
<pre><code>name: Run Tests
on: pull_request
jobs:
  test:
    runs-on: ubuntu-latest
    services:
      mysql:
        image: mysql:5.7
        ports:
        - 3306
        options: --health-cmd "mysqladmin ping" --health-interval 10s --health-timeout 5s --health-retries 10

    steps:
      - uses: actions/checkout@v1.0.0
        with:
          fetch-depth: 1
      - uses: actions/setup-elixir@v1.0.0
        with:
          otp-version: 22.x
          elixir-version: 1.9.x
      - run: MIX_ENV=test mix deps.get
      - name: setup db and test
        run: |
          MIX_ENV=test mix ecto.setup
          mix test
        env:
          DATABASE_URL: "mysql2://root:root@127.0.0.1:${{ job.services.mysql.ports['3306'] }}/myapp"
</code></pre>
<p>That config was working perfectly fine for a week or two and then, seemingly out of nowhere, it started failing consistently with the service reporting (almost immediately) that mysql is unhealthy:</p>
<pre><code>Status: Downloaded newer image for mysql:5.7
docker.io/library/mysql:5.7
/usr/bin/docker create --name 926a41768ca34ea8b1364385026a6af1_mysql57_8b5743 --label 2e4e3a --network github_network_169d64eaca1a440991d3382af497b76f --network-alias mysql -p 3306 --health-cmd "mysqladmin ping" --health-interval 10s --health-timeout 5s --health-retries 10 -e GITHUB_ACTIONS=true mysql:5.7
03afaad28cc629cb34a9507a2d19435c2c32ddc1f7fd4f94da2033c01f26c938
/usr/bin/docker start 03afaad28cc629cb34a9507a2d19435c2c32ddc1f7fd4f94da2033c01f26c938
03afaad28cc629cb34a9507a2d19435c2c32ddc1f7fd4f94da2033c01f26c938
/usr/bin/docker ps --all --filter id=03afaad28cc629cb34a9507a2d19435c2c32ddc1f7fd4f94da2033c01f26c938 --filter status=running --no-trunc --format "{{.ID}} {{.Status}}"
03afaad28cc629cb34a9507a2d19435c2c32ddc1f7fd4f94da2033c01f26c938 Up Less than a second (health: starting)
/usr/bin/docker port 03afaad28cc629cb34a9507a2d19435c2c32ddc1f7fd4f94da2033c01f26c938
/usr/bin/docker inspect --format="{{if .Config.Healthcheck}}{{print .State.Health.Status}}{{end}}" 03afaad28cc629cb34a9507a2d19435c2c32ddc1f7fd4f94da2033c01f26c938
unhealthy
##[error]Failed to initialize, mysql service is unhealthy.
</code></pre>
<p>As I mentioned, that final error (<code>Failed to initialize, mysql service is unhealthy.</code>) appears almost immediately, and it seems that my health cmd and interval options aren’t being respected.</p>
<p>Anyone know what I might be missing to get this to work?</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi peburrows,</p>
<p>Glad to hear you in Github Community Forum!</p>
<p>I copied your code and reproduced the issue. After i add ‘<a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fhelp.github.com%2Fen%2Factions%2Fautomating-your-workflow-with-github-actions%2Fworkflow-syntax-for-github-actions%23jobsjob_idservicesenv&amp;data=02%7C01%7Cv-wadzho%40microsoft.com%7C3a28099b7f404e8f96f408d7698d9ad3%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637093929214576023&amp;sdata=hCoqPa7D4wp0clp635R2FphulFrtSMHNQ7iHsV%2F55VM%3D&amp;reserved=0" rel="nofollow noopener">jobs.&lt;job_id&gt;.services.env</a>’ to the service, the error is gone, code as below, hope it helps, thanks.</p>
<pre><code>services:
      mysql:
        image: mysql:5.7
        env:
          MYSQL_ROOT_PASSWORD: password
        ports:
        - 3306
        options: --health-cmd "mysqladmin ping" --health-interval 10s --health-timeout 5s --health-retries 10
</code></pre>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/4/4fb839462535f6e5b22968eb6373c2f069f555bd.png" data-download-href="/uploads/short-url/bnepTyGtEjbCDPZu5bgDIEJmU0Z.png?dl=1" title="MySQL.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181101174-05c6099c-5a81-4417-bfdc-f2565e61a01f.png" alt="MySQL.png" data-base62-sha1="bnepTyGtEjbCDPZu5bgDIEJmU0Z" width="690" height="308" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/4/4fb839462535f6e5b22968eb6373c2f069f555bd_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">MySQL.png</span><span class="informations">1085×485 30.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by peburrows** [comment]
<p>Awesome, thank you! That seems to have fixed it.</p>

**Comment by macgill4444** [comment]
<p>Just for future reference for others that might run into this issue. I had an action that setup a postgres and a redis service. This was what the service code looked like in my main.yml:</p>
<pre><code>services:
db:
image: postgres:11
ports: ['5432:5432']
options: &gt;-
--health-cmd pg_isready
--health-interval 10s
--health-timeout 5s
--health-retries 5
redis:
image: redis
ports: ['6379:6379']
options: --entrypoint redis-server
</code></pre>
<p>It had been running completely fine for a few weeks and then out of the blue the setup of the redis server start failing with the roughly the same issue.</p>
<pre><code>##[error]Failed to initialize, db service is unhealthy.
</code></pre>
<p>I came across this post and thought it wasn’t relevant at first. After much testing I tried setting up my redis service through another Action from the marketplace. At that point my **postgres** service started failing with the same message above!</p>
<pre><code>##[error]Failed to initialize, db service is unhealthy.
</code></pre>
<p>Roughly following the advice from this thread I added an .env section with passwords, username, and db env variables to my postgres service. It worked! This is what the code looks like now: </p>
<pre><code>services:
db:
image: postgres:11
env:
POSTGRES\_USER: postgres
POSTGRES\_PASSWORD: postgres
POSTGRES\_DB: postgres
ports: ['5432:5432']
options: &gt;-
--health-cmd pg_isready
--health-interval 10s
--health-timeout 5s
--health-retries 5
redis:
image: redis
ports: ['6379:6379']
options: --entrypoint redis-server
</code></pre>
<p>Hope this is helpful for future devs. Also no clue why this suddenly started happening. Nothing changed in my code that would have affected this.</p>

**Comment by phiat** [comment]
<p>I just had same issue with postgres 11, running fine for weeks then today:  “Failed to initialize, db service is unhealthy”.  Just started about an hour ago.     This is only thing I found on web that discusses this.   Adding the env part worked first time!  Thank you very much!  </p>

**Comment by mattlima** [comment]
<p>Same here, my postgres container stopped working, this fixed it. Thanks!</p>

**Comment by sampathaya** [comment]
<p>Hey can you show the changes you have done on script? thanks</p>

**Comment by ahaverty** [comment]
<p>That fixed the container step, but in my rails db:prepare step, I was suddenly getting an issue about a missing postgres password.</p>
<p>Make sure to include the env’s:</p>
<pre><code>DATABASE_URL: postgres://postgres:@localhost:5432/test
          PGUSER: postgres
          PGPASSWORD: postgres
</code></pre>
<p>for any other step that needs postgres, such as:</p>
<pre><code>bundle exec rails db:prepare
</code></pre>

**Comment by ProbablyFaiz** [comment]
<p>Thank you! This resolved my issue.</p>

**Comment by lastobelus** [comment]
<p>It seems that github actions no longer passes the job:env to services, and that you have to replicate the service env vars.</p>
<p>ie, this used to work:</p>
<pre><code>build:
    runs-on: ubuntu-latest
    env:
      POSTGRES_USER: postgres
      POSTGRES_PASSWORD: postgres
      POSTGRES_DB: postgres

    services:
      postgres:
        image: postgres:11-alpine
        ports:
          - 5432:5432
        options: &gt;-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
</code></pre>
<p>but now you have to repeat the env for the service:</p>
<pre><code>build:
    runs-on: ubuntu-latest
    env:
      POSTGRES_USER: postgres
      POSTGRES_PASSWORD: postgres
      POSTGRES_DB: postgres

    services:
      postgres:
        image: postgres:11-alpine
        env:
          POSTGRES_USER: postgres
          POSTGRES_PASSWORD: postgres
          POSTGRES_DB: postgres
        ports:
          - 5432:5432
        options: &gt;-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
</code></pre>
<p>Github: stop making it hard to DRY!!!</p>

**Comment by yukster** [comment]
<p>Another happy postgres user whose CI action mysteriously stopped working and found this thread. Thanks for getting me running again!</p>

**Comment by baspeeters** [comment]
<p>I’d tried all of the suggestions here, but still received the same error.</p>
<p>Maybe it had something to do with me using mysql:5.6 instead of a higher version like I see being used in every snippet.</p>
<p>Anyway, after moving the environment variables to the service, using the MySQL server image did the trick for me:</p>
<pre><code>services:
      mysql:
        image: mysql/mysql-server:5.6
        env:
          MYSQL_USER: testuser
          MYSQL_PASSWORD: testpassword
        options: &gt;-
          --health-cmd "mysql health ping"
          --health-interval 10s
          --health-timeout 10s
          --health-retries 10
</code></pre>

**Comment by joeyagreco-sb** [comment]
Here's what worked for me:

- Check if healthy
- If not healthy, log most recent docker logs

```yaml
      - name: Check MySQL health status
        id: check_health
        run: |
          echo "Checking MySQL health status..."
          attempt_counter=0
          max_attempts=24
          until [ "$(docker inspect --format="{{.State.Health.Status}}" my_container)" == "healthy" ]; do
            if [ ${attempt_counter} -eq ${max_attempts} ];then
              echo "Max attempts reached"
              exit 1
            fi

            health_status=$(docker inspect --format="{{.State.Health.Status}}" my_container)
            echo "Current MySQL health status: ${health_status}"

            if [ "${health_status}" != "healthy" ]; then
              echo "Displaying last 10 lines from Docker logs:"
              docker logs --tail 10 my_container
            fi

            attempt_counter=$(($attempt_counter+1))
            sleep 5
          done
```
